### PR TITLE
fix: fix `$bind` and `:repl` being replaced in user input

### PR DESCRIPTION
**Pull request by ephys**
This PR is still WIP but I'm opening it to describe the changes as I go.

Fixes in v7: #13817 

This (v7) PR fixes how bind & replacements are handled in query interface & findAll methods.

### Before this PR

Let's take the following query:

```typescript
User.findAll({
  where: and(
    literal('soundex("firstName") = soundex(:firstName)'),
    { lastName: lastName },
  ),
  replacements: { firstName },
})
```

Before this PR, Sequelize would first generate the SQl for the `where` parameter:

```sql
SELECT * FROM users WHERE soundex("firstName") = soundex(:firstName) AND "lastName" = 'Doe'
```

Then pass that query to `Sequelize#query`, which would process the `replacements` option (in this case, the `:firstName` parameter):

```sql
SELECT * FROM users WHERE soundex("firstName") = soundex('John') AND "lastName" = 'Doe'
```

---

Because of this two-step processing, a malicious input can result in SQL injection. For instance with these two user inputs:

```js
const lastName = ':firstName'; // <- notice how it includes the syntax for a replacement parameter
const firstName = '; DROP TABLE users;'
```

With the above input, `QueryGenerator` will first generate this:

```sql
SELECT * FROM users WHERE soundex("firstName") = soundex(:firstName) AND "lastName" = ':firstName'
```

Then `Sequelize#query` will process `:firstName`, resulting in this:

```sql
SELECT * FROM users WHERE soundex("firstName") = soundex('; DROP TABLE users;') AND "lastName" = ''; DROP TABLE users;''
```

Resulting in the injection of `DROP TABLE users;`, interpreted as actual SQL.

### After this PR

Replacements are now handled by `QueryGenerator` instead. This allows `QueryGenerator` to *only* process replacements in `literal()` and nowhere else.

`QueryGenerator` will generate this:

```sql
SELECT * FROM users WHERE soundex("firstName") = soundex('; DROP TABLE users;') AND "lastName" = ':firstName'
```

Then the query will be passed to a new method, `Sequelize#queryRaw`, which does *not* process replacements at all. So the query is executed as-is.

`Sequelize#query` will still process replacements, but as long as you don't put replacements inside strings there, and don't inline user input yourself, you'll be fine.

### Other changes

- The `bind` & `replacement` options can now be used at the same time! This is because handling of bind parameters is done by a new parser that knows whether `$name` is a bind parameter, or just text inside a string.
- It's now possible to use `bind` parameters in `update`, and `insert` methods. It was not previously possible because sequelize generates its own bind parameters in these methods, which would overwrite the ones provided by the user. They are now safely merged.

### TODO

- Test the new `bind` replacer against comments

### CODE REVIEW

**Comment by ephys** [comment]
Crossing my fingers for the last test suite but i think this may be ready

**Comment by ephys** [comment]
Breaking changes documented https://github.com/sequelize/website/pull/135

File: test/unit/query-generator/select-query.test.ts [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+      ibmi: 'SELECT "id" FROM "Users" AS "User" OFFSET 1 ROWS',
+      mssql: `SELECT [id] FROM [Users] AS [User] ORDER BY [User].[id] OFFSET 1 ROWS;`,
+    });
+  });
+
+  describe('replacements', () => {
+    it('parses named replacements in literals', async () => {
+      // The goal of this test is to test that :replacements are parsed in literals in as many places as possible
+
+      const sql = queryGenerator.selectQuery(User.tableName, {
```
Comment (WikiRik) [inline test/unit/query-generator/select-query.test.ts:54]: What a test, good work!

File: test/unit/query-generator/select-query.test.ts [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+            INNER JOIN "Users" AS "contributors"
+            ON "contributors"."id" = "contributors->ProjectContributor"."UserId"
+            AND 'where'
+          )
+          ON "Project"."id" = "contributors->ProjectContributor"."ProjectId"
+        `,
+      });
+    });
+
+    // TODO: test subQueries
```
Comment (WikiRik) [inline test/unit/query-generator/select-query.test.ts:270]: This is for a future PR, right?

File: test/unit/query-interface/decrement.test.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+          id: 2,
+          data: 3,
+        },
+      },
+    );
+
+    expect(stub.callCount).to.eq(1);
+    const firstCall = stub.getCall(0);
+    expectsql(firstCall.args[0] as string, {
+      postgres: `UPDATE "Users" SET "age"="age"- ':age',"name"=':name' WHERE "id" = ':id' RETURNING ":data"`,
```
Comment (WikiRik) [inline test/unit/query-interface/decrement.test.ts:42]: Not sure how I can get a comment on multiple lines, but can you swap `postgres` here with `default`?

File: test/unit/query-interface/increment.test.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+          id: 2,
+          data: 3,
+        },
+      },
+    );
+
+    expect(stub.callCount).to.eq(1);
+    const firstCall = stub.getCall(0);
+    expectsql(firstCall.args[0] as string, {
+      postgres: `UPDATE "Users" SET "age"="age"+ ':age',"name"=':name' WHERE "id" = ':id' RETURNING ":data"`,
```
Comment (WikiRik) [inline test/unit/query-interface/increment.test.ts:42]: Same comment as above, can you swap `postgres` and `default`?

File: test/unit/query-interface/insert.test.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+      mssql: `INSERT INTO [Users] ([firstName]) OUTPUT INSERTED.[:data] VALUES ($sequelize_1);`,
+      db2: `SELECT * FROM FINAL TABLE (INSERT INTO "Users" ("firstName") VALUES ($sequelize_1));`,
+      ibmi: `SELECT * FROM FINAL TABLE (INSERT INTO "Users" ("firstName") VALUES ($sequelize_1))`,
+    });
+    expect(firstCall.args[1]?.bind).to.deep.eq({
+      sequelize_1: 'Zoe',
+    });
+  });
+
+  it('throws if a bind parameter name starts with the reserved "sequelize_" prefix', async () => {
```
Comment (WikiRik) [inline test/unit/query-interface/insert.test.ts:42]: Now that I think about this, why is this test only present in some files?

File: test/integration/sequelize/query.test.js [hunk 6]
```diff
@@ -807,29 +733,33 @@ describe(Support.getTestDialectTeaser('Sequelize'), () => {
...
-        const expected = ['db2', 'ibmi'].includes(dialect) ? [{ FOO: 1, BAR: '$ / $1' }] : [{ foo: 1, bar: '$ / $1' }];
+        const expected = ['db2', 'ibmi'].includes(dialect) ? [{ FOO: 1, BAR: '$$ / $$1' }] : [{ foo: 1, bar: '$$ / $$1' }];
         expect(result[0]).to.deep.equal(expected);
         if (['postgres', 'sqlite', 'db2', 'ibmi'].includes(dialect)) {
           expect(logSql).to.include('$1');
         }
       });
 
-      it('binds named parameters object handles escaped $$', async function () {
+      // this was a legacy band aid that has since been removed, because the underlying issue (transforming bind params in strings) has been fixed.
```
Comment (WikiRik) [inline test/integration/sequelize/query.test.js:754]: Like above, do we need this comment? I get that it's useful for the review but doesn't need to stay here right?

File: test/integration/sequelize/query.test.js [hunk 7]
```diff
@@ -807,29 +733,33 @@ describe(Support.getTestDialectTeaser('Sequelize'), () => {
         expect(result[0]).to.deep.equal(expected);
       });
 
-      it('binds named parameters array handles escaped $$', async function () {
+      // this was a legacy band aid that has since been removed, because the underlying issue (transforming bind params in strings) has been fixed.
```
Comment (WikiRik) [inline test/integration/sequelize/query.test.js:736]: Do we need this comment?

**Review by WikiRik — COMMENTED** [review]
Finished looking at the tests, do have a few comments but looks great! Would be nice if we could clear up most of the TODOs before releasing v7

File: test/integration/model.test.js [hunk 8]
```diff
@@ -1054,11 +1056,18 @@ describe(Support.getTestDialectTeaser('Model'), () => {
...
-          if (['mssql', 'ibmi'].includes(dialect)) {
-            expect(sql).to.not.contain('createdAt');
-          } else {
-            expect(sql).to.match(/UPDATE\s+["`]+User1s["`]+\s+SET\s+["`]+secretValue["`]=(\$1|\?),["`]+updatedAt["`]+=(\$2|\?)\s+WHERE ["`]+id["`]+\s=\s(\$3|\?)/);
-          }
+
+          expect(sql).to.match(/^Executing \(default\): /);
+          sql = sql.slice(21);
+
+          expectsql(sql, {
```
[the remark this answers is not in the document]
Comment (ephys) [reply test/integration/model.test.js:1063]: It's probably similar to db2, but I don't really know for sure as I can't run this code against ibmi

File: test/integration/sequelize/query.test.js (hunk 7 above)
Comment (ephys) [reply test/integration/sequelize/query.test.js:736]: It's there to explain why a test that does seemingly nothing exists as not transforming the contents of strings seems like the logical thing to do (at least to me) and someone may be tempted to remove it, but I kept the test because we used to have a weird behavior there

File: test/unit/query-interface/insert.test.ts (hunk 5 above)
Comment (ephys) [reply test/unit/query-interface/insert.test.ts:42]: They're the only methods reserving some bind parameter names, because they're the only methods that add extra bind parameters

File: test/unit/query-interface/decrement.test.ts (hunk 3 above)
Comment (ephys) [reply test/unit/query-interface/decrement.test.ts:42]: click on one line, then shift+click on another one to comment on multiple lines :)

File: test/unit/query-generator/select-query.test.ts (hunk 2 above)
Comment (ephys) [reply test/unit/query-generator/select-query.test.ts:270]: Ideally no, I should add these here

File: test/unit/query-generator/bulk-insert-query.test.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
+import { DataTypes, literal } from '@sequelize/core';
+import { expectsql, sequelize } from '../../support';
+
+describe('QueryGenerator#bulkInsertQuery', () => {
+  const queryGenerator = sequelize.getQueryInterface().queryGenerator;
+
+  const User = sequelize.define('User', {
+    firstName: DataTypes.STRING,
+  }, { timestamps: false });
+
```
[the remark this answers is not in the document]
Comment (ephys) [reply test/unit/query-generator/bulk-insert-query.test.ts:10]: Yep, would be good to add these in each query-interface file if the query is supposed to support replacements

File: test/unit/query-generator/bulk-insert-query.test.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    }], {
+      replacements: {
+        injection: 'a string',
+      },
+    });
+
+    expectsql(sql, {
+      default: `INSERT INTO [Users] ([firstName]) VALUES ('a string');`,
+      mssql: `INSERT INTO [Users] ([firstName]) VALUES (N'a string');`,
+      // TODO: db2 should use the same system as ibmi
```
[the remark this answers is not in the document]
Comment (ephys) [reply test/unit/query-generator/bulk-insert-query.test.ts:23]: I may just remove the comment. I think `ibmi` does something different 'incorrectly'. I think its query is written like this to emulate 'RETURNING', but the option is not enabled, so `ibmi` may be wrong here

File: src/dialects/abstract/query-generator.js [hunk 11]
```diff
@@ -2254,7 +2268,20 @@ export class AbstractQueryGenerator {
     }
 
     if (smth instanceof Utils.Literal) {
-      return smth.val;
+      if (includesPositionalReplacements(smth.val)) {
+        throw new TypeError(`The following literal includes positional replacements (?). Only named replacements (:name) are allowed in literal() because we cannot guarantee the order in which they will be evaluated:
```
Comment (WikiRik) [inline src/dialects/abstract/query-generator.js:2272]: Can we format this so that it's more readable?

Comment (WikiRik) [reply src/dialects/abstract/query-generator.js:2272]: As in; that it doesn't take up the entire screen

File: src/dialects/abstract/query-interface.js [hunk 12]
```diff
@@ -1066,10 +1119,14 @@ export class QueryInterface {
 
     const dataType = options.dataType;
 
+    // FIXME: DECIMAL is not safely representable as a float!
```
Comment (WikiRik) [inline src/dialects/abstract/query-interface.js:1122]: This is for a next PR right? And why did you use FIXME here instead of TODO?

**Review by WikiRik — COMMENTED** [review]
Few additional comments, but again; it looks great. Thanks for your hard work on this!

File: test/integration/model.test.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (WikiRik) [reply test/integration/model.test.js:1063]: So we should add a TODO to check this, right?

File: test/integration/sequelize/query.test.js (hunk 7 above)
Comment (WikiRik) [reply test/integration/sequelize/query.test.js:736]: Ah yes, that's a good reason

File: test/unit/utils/utils.test.ts [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+      it('can take a value to compare against', () => {
+        const path = 'metadata.pg_rating.is';
+        const value = 'U';
+        expect(queryGenerator.handleSequelizeMethod(new Utils.Json(path, value))).to.equal('("metadata"#>>\'{pg_rating,is}\') = \'U\'');
+      });
+    });
+  }
+
+  describe('inflection', () => {
+    it('works better than lingo ;)', () => {
```
[the remark this answers is not in the document]
Comment (fzn0x) [reply test/unit/utils/utils.test.ts:132]: Haha this one has a funny mesage

File: src/dialects/mssql/index.js [hunk 14]
```diff
@@ -49,6 +51,11 @@ export class MssqlDialect extends AbstractDialect {
       this.queryGenerator,
     );
   }
+
+  createBindCollector() {
+    // unlike postgres, sqlite supports $named bind parameters, but it wasn't deemed worth it to implement it.
```
[the remark this answers is not in the document]
Comment (ephys) [reply src/dialects/mssql/index.js:56]: Not only is it mssql, but I ended up implementing a collector that handles named parameters for mssql & sqlite anyway

File: src/dialects/mssql/query-interface.js [hunk 15]
```diff
@@ -84,6 +90,8 @@ export class MsSqlQueryInterface extends QueryInterface {
 
     const sql = this.queryGenerator.upsertQuery(tableName, insertValues, updateValues, where, model, options);
 
-    return await this.sequelize.query(sql, options);
+    delete options.replacements;
```
[the remark this answers is not in the document]
Comment (ephys) [reply src/dialects/mssql/query-interface.js:93]: `queryRaw` will throw if `options.replacements` is specified in case someone tries to use it like `query`
I delete it before calling queryRaw to prevent it from throwing. I thought I documented it each time but it's missing here

File: src/sequelize.d.ts [hunk 16]
```diff
@@ -1265,6 +1366,33 @@ export class Sequelize extends Hooks {
   public query(sql: string | { query: string; values: unknown[] }, options: (QueryOptions | QueryOptionsWithType<QueryTypes.RAW>) & { plain: true }): Promise<{ [key: string]: unknown } | null>;
   public query(sql: string | { query: string; values: unknown[] }, options?: QueryOptions | QueryOptionsWithType<QueryTypes.RAW>): Promise<[unknown[], unknown]>;
 
+  /**
+   * Works like {@link Sequelize#query}, but does not inline replacements. Only bind parameters are supported.
+   *
+   * @param sql The SQL to execute
+   * @param options The options for the query. See {@link QueryRawOptions} for details.
+   */
+  public queryRaw(sql: string | { query: string; values: unknown[] }, options: QueryRawOptionsWithType<QueryTypes.UPDATE>): Promise<[undefined, number]>;
```
[the remark this answers is not in the document]
Comment (ephys) [reply src/sequelize.d.ts:1375]: They're basically the same as `query()`, but they do not accept the `replacements` option

File: src/sequelize.js [hunk 17]
```diff
@@ -534,10 +536,70 @@ export class Sequelize {
...
+  }
+
+  async queryRaw(sql, options) {
+    if (typeof sql !== 'string') {
+      throw new TypeError('Sequelize#rawQuery requires a string as the first parameter.');
+    }
+
+    if (options != null && 'replacements' in options) {
+      throw new TypeError(`Sequelize#rawQuery does not accept the "replacements" options.
+Only bind parameters can be provided, in the dialect-specific syntax.
```
[the remark this answers is not in the document]
Comment (ephys) [reply src/sequelize.js:563]: It's on purpose, if I indent the string, it'll end up having a lot of whitespace. This is something https://github.com/tc39/proposal-string-dedent will improve :D

File: src/dialects/abstract/query-interface.js (hunk 12 above)
Comment (ephys) [reply src/dialects/abstract/query-interface.js:1122]: For another PR yep, it's just a thing I noticed while I was working on that file
I'll change it to TODO for consistency. I sometimes use `FIXME` when it's a bug but it's probably a bad idea, `TODO` is universally recognized, `FIXME` isn't

File: src/sequelize.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (WikiRik) [reply src/sequelize.js:563]: Ah yes

File: test/integration/model.test.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (ephys) [reply test/integration/model.test.js:1063]: Yep!

[the remark this answers is not in the document]
Comment (ephys) [reply test/integration/model.test.js:1063]: its output is likely to be `UPDATE "users1" SET "secretValue"=?,"updatedAt"=? WHERE "id" = ?;`, I'll add that

File: test/unit/query-generator/bulk-insert-query.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (ephys) [reply test/unit/query-generator/bulk-insert-query.test.ts:10]: Oh, this is already a query-generator test actually

File: test/unit/utils/utils.test.ts [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+      expect(Utils.pluralize('status')).to.equal('statuses');
+
+      expect(Utils.singularize('status')).to.equal('status');
+    });
+  });
+
+  describe('flattenObjectDeep', () => {
+    it('should return the value if it is not an object', () => {
+      const value = 'non-object';
+      // @ts-expect-error
```
[the remark this answers is not in the document]
Comment (ephys) [reply test/unit/utils/utils.test.ts:145]: I've changed the signature of `flattenObjectDeep` to accept any input since that's what it does

File: test/unit/dialects/abstract/query-generator.test.js [hunk 19]
```diff
@@ -122,10 +123,15 @@ describe('QueryGenerator', () => {
         .should.be.equal('foo IS NOT NULL');
     });
 
-    it('should correctly escape $ in sequelize.fn arguments', function () {
+    // this was a band-aid over a deeper problem ('$bind' being considered to be a bind parameter when it's a string), which has been fixed
```
[the remark this answers is not in the document]
Comment (ephys) [reply test/unit/dialects/abstract/query-generator.test.js:126]: I prefer to keep it to explain why there is a test that checks that something is not done

File: test/unit/query-interface/decrement.test.ts (hunk 3 above)
Comment (ephys) [reply test/unit/query-interface/decrement.test.ts:42]: click one line, then shift+click another to comment on multiple lines :)

File: test/unit/query-interface/insert.test.ts (hunk 5 above)
Comment (ephys) [reply test/unit/query-interface/insert.test.ts:42]: Only insert, update, and bulkUpdate add extra bind parameters in their SQL, so they're the only 3 methods that reserve bind parameters

**Comment by ephys** [comment]
Changes implemented. I just need to add this test now:

> // TODO: test subQueries

File: src/dialects/abstract/query-generator.js [hunk 20]
```diff
@@ -469,9 +471,13 @@ export class AbstractQueryGenerator {
    * @private
    */
   arithmeticQuery(operator, tableName, where, incrementAmountsByField, extraAttributesToBeUpdated, options) {
+    // TODO: this method should delegate to `updateQuery`
```
[the remark this answers is not in the document]
Comment (WikiRik) [reply src/dialects/abstract/query-generator.js:474]: No worries, we can leave it like this

**Review by WikiRik — DISMISSED** [review]
New test looks good and other comments have also been addressed so I think this is good to go

**Comment by ephys** [comment]
There is one last fix I want to add before merging.
Now that we have a string-aware SQL "parser", we can prevent replacements from being transformed in literals too `literal('= \':value\'')` should not consider `:value` to be a replacement

File: test/unit/utils/sql.test.ts [hunk 21]
```diff
@@ -264,3 +264,327 @@ describe('mapBindParameters', () => {
...
+    }).to.throw('Named replacement ":toString" has no entry in the replacement map.');
+
+  });
+
+  it('parses named replacements followed by cast syntax', () => {
+    const sql = injectReplacements(`SELECT * FROM users WHERE id = :id::string`, dialect, {
+      id: 1,
+    });
+
+    // this is postgres-only syntax. It doesn't work in other dialects.
```
Comment (WikiRik) [inline test/unit/utils/sql.test.ts:293]: Do we want to run this test on all dialects then?

File: test/unit/utils/sql.test.ts [hunk 22]
```diff
@@ -264,3 +264,327 @@ describe('mapBindParameters', () => {
...
 });
+
+describe('injectReplacements (named replacements)', () => {
+  it('parses named replacements', () => {
+    const sql = injectReplacements(`SELECT ${dialect.TICK_CHAR_LEFT}:id${dialect.TICK_CHAR_RIGHT} FROM users WHERE id = ':id' OR id = :id OR id = ''':id'''`, dialect, {
+      id: 1,
+    });
+
+    expectsql(sql, {
+      default: `SELECT [:id] FROM users WHERE id = ':id' OR id = 1 OR id = ''':id'''`,
```
Comment (WikiRik) [inline test/unit/utils/sql.test.ts:275]: Why are `dialect.TICK_CHAR_LEFT` and `dialect.TICK_CHAR_RIGHT` not needed in the result?

File: test/unit/utils/sql.test.ts [hunk 23]
```diff
@@ -264,3 +264,327 @@ describe('mapBindParameters', () => {
...
+      a: 1,
+    });
+
+    expectsql(sql, {
+      default: `SELECT * FROM users WHERE id = 1`,
+    });
+  });
+
+  it(`does not consider the token to be a replacement if it does not follow '(', ',', '=' or whitespace`, () => {
+    const sql = injectReplacements(`SELECT * FROM users WHERE id = fn(:id) OR id = fn('a',:id) OR id=:id`, dialect, {
```
Comment (WikiRik) [inline test/unit/utils/sql.test.ts:310]: Do we want to add `OR id = :id` here because the test name mentions whitespace?

**Review by WikiRik — COMMENTED** [review]
In general looks good, I have three comments about the tests. Same comments also apply to their respective equivalent for positional replacements

File: test/unit/utils/sql.test.ts [hunk 24]
```diff
@@ -0,0 +1,590 @@
...
+        SELECT * FROM users /*
+        WHERE id = ?
+        */
+        WHERE id = 1
+      `,
+    });
+  });
+
+  // https://github.com/sequelize/sequelize/issues/14358
+  it('does not parse ?& and ?| operators as replacements (#14358)', async () => {
```
Comment (WikiRik) [inline test/unit/utils/sql.test.ts:551]: Do we want to test this one on all dialects even though it's postgres exclusive?

File: test/unit/utils/sql.test.ts (hunk 22 above)
Comment (ephys) [reply test/unit/utils/sql.test.ts:275]: `default` automatically converts `[` and `]` into the right tick for the dialect being tested (except in the case of `ARRAY[x]`)

File: test/unit/utils/sql.test.ts (hunk 23 above)
Comment (ephys) [reply test/unit/utils/sql.test.ts:310]: Ah yes

File: test/unit/utils/sql.test.ts (hunk 24 above)
Comment (ephys) [reply test/unit/utils/sql.test.ts:551]: Yep! I'd prefer it to be consistent across all dialects, it could be added to other dialects in the future if they start supporting arrays (last I heard arrays were added to the SQL spec)

File: test/unit/utils/sql.test.ts (hunk 22 above)
Comment (WikiRik) [reply test/unit/utils/sql.test.ts:275]: Ah, that makes sense. Thanks!

File: test/unit/utils/sql.test.ts (hunk 23 above)
Comment (ephys) [reply test/unit/utils/sql.test.ts]: Done

File: test/unit/utils/sql.test.ts (hunk 21 above)
Comment (ephys) [reply test/unit/utils/sql.test.ts:293]: Turns out more dialects support it. It's not unlikely that it may get adopted by even more some day, might as well make sure we parse it properly

**Review by WikiRik — APPROVED** [review]
Thanks again for your thorough work on this!

**Comment by ephys** [comment]
Now onto v6
